File: packages/snow_draw_engine/lib/draw/elements/types/arrow/core/arrow_order_core.dart
import 'arrow_hit_test.dart';
import 'arrow_types.dart';

ReorderArrowAboveElementsResult _createUnchangedResult(
  List<String> orderedElementIds,
) => ReorderArrowAboveElementsResult(
  orderedElementIds: orderedElementIds,
  moved: false,
  fromIndex: -1,
  toIndex: -1,
);

ReorderArrowAboveHoveredBindableResult _createHoveredUnchangedResult(
  List<String> orderedElementIds, {
  String? hoveredBindableId,
  List<String> anchorElementIds = const <String>[],
}) => ReorderArrowAboveHoveredBindableResult(
  orderedElementIds: orderedElementIds,
  moved: false,
  fromIndex: -1,
  toIndex: -1,
  hoveredBindableId: hoveredBindableId,
  anchorElementIds: anchorElementIds,
);

Map<String, List<String>> _toAnchorLookup(AnchorElementIdsLookupInput? input) {
  if (input == null) {
    return <String, List<String>>{};
  }
  if (input is Map<String, List<String>>) {
    return input;
  }
  if (input is Map) {
    return input.map<String, List<String>>(
      (key, value) => MapEntry(key as String, (value as List).cast<String>()),
    );
  }
  return <String, List<String>>{};
}

ReorderArrowAboveElementsResult reorderArrowAboveElements(
  ReorderArrowAboveElementsInput input,
) {
  if (input.orderedElementIds.isEmpty || input.anchorElementIds.isEmpty) {
    return _createUnchangedResult(input.orderedElementIds);
  }

  final fromIndex = input.orderedElementIds.indexOf(input.arrowId);
  if (fromIndex == -1) {
    return _createUnchangedResult(input.orderedElementIds);
  }

  final anchorIdSet = input.anchorElementIds.toSet();
  final toIndex = input.orderedElementIds.indexWhere(anchorIdSet.contains);
  if (toIndex == -1 || fromIndex >= toIndex) {
    return _createUnchangedResult(input.orderedElementIds);
  }

  final nextOrderedElementIds = List<String>.from(input.orderedElementIds);
  final arrow = nextOrderedElementIds.removeAt(fromIndex);
  nextOrderedElementIds.insert(toIndex, arrow);

  return ReorderArrowAboveElementsResult(
    orderedElementIds: nextOrderedElementIds,
    moved: true,
    fromIndex: fromIndex,
    toIndex: toIndex,
  );
}

ReorderArrowAboveHoveredBindableResult reorderArrowAboveHoveredBindable(
  ReorderArrowAboveHoveredBindableInput input,
) {
  String? hoveredBindableId;
  if (input.hoveredBindableId != null && input.hoveredBindableId!.isNotEmpty) {
    hoveredBindableId = input.hoveredBindableId;
  } else if (input.point != null && input.bindables != null) {
    hoveredBindableId = getHoveredBindable(
      input.point!,
      input.bindables!,
      input.tolerance ?? 0,
    )?.id;
  }

  if (hoveredBindableId == null) {
    return _createHoveredUnchangedResult(input.orderedElementIds);
  }

  final anchorLookup = _toAnchorLookup(input.anchorElementIdsByBindableId);
  final anchorElementIds =
      anchorLookup[hoveredBindableId] ?? <String>[hoveredBindableId];

  final reorder = reorderArrowAboveElements(
    ReorderArrowAboveElementsInput(
      orderedElementIds: input.orderedElementIds,
      arrowId: input.arrowId,
      anchorElementIds: anchorElementIds,
    ),
  );

  return ReorderArrowAboveHoveredBindableResult(
    orderedElementIds: reorder.orderedElementIds,
    moved: reorder.moved,
    fromIndex: reorder.fromIndex,
    toIndex: reorder.toIndex,
    hoveredBindableId: hoveredBindableId,
    anchorElementIds: anchorElementIds,
  );
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/core/arrow_order_core.rs
#![allow(dead_code)]

use std::collections::HashMap;

use super::arrow_hit_test::get_hovered_bindable;
use super::arrow_types::{
    BindableState, ReorderArrowAboveElementsInput, ReorderArrowAboveElementsResult,
    ReorderArrowAboveHoveredBindableInput, ReorderArrowAboveHoveredBindableResult,
};

fn unchanged_result(ordered_element_ids: &[String]) -> ReorderArrowAboveElementsResult {
    ReorderArrowAboveElementsResult {
        ordered_element_ids: ordered_element_ids.to_vec(),
        moved: false,
        from_index: -1,
        to_index: -1,
    }
}

fn hovered_unchanged_result(
    ordered_element_ids: &[String],
    hovered_bindable_id: Option<String>,
    anchor_element_ids: Vec<String>,
) -> ReorderArrowAboveHoveredBindableResult {
    ReorderArrowAboveHoveredBindableResult {
        ordered_element_ids: ordered_element_ids.to_vec(),
        moved: false,
        from_index: -1,
        to_index: -1,
        hovered_bindable_id,
        anchor_element_ids,
    }
}

pub fn reorder_arrow_above_elements(
    input: &ReorderArrowAboveElementsInput,
) -> ReorderArrowAboveElementsResult {
    if input.ordered_element_ids.is_empty() || input.anchor_element_ids.is_empty() {
        return unchanged_result(&input.ordered_element_ids);
    }

    let Some(from_index) = input
        .ordered_element_ids
        .iter()
        .position(|id| id == &input.arrow_id)
    else {
        return unchanged_result(&input.ordered_element_ids);
    };

    let Some(to_index) = input
        .ordered_element_ids
        .iter()
        .position(|id| input.anchor_element_ids.iter().any(|anchor| anchor == id))
    else {
        return unchanged_result(&input.ordered_element_ids);
    };

    if from_index >= to_index {
        return unchanged_result(&input.ordered_element_ids);
    }

    let mut ordered = input.ordered_element_ids.clone();
    let arrow_id = ordered.remove(from_index);
    ordered.insert(to_index, arrow_id);

    ReorderArrowAboveElementsResult {
        ordered_element_ids: ordered,
        moved: true,
        from_index: from_index as isize,
        to_index: to_index as isize,
    }
}

pub fn reorder_arrow_above_hovered_bindable(
    input: &ReorderArrowAboveHoveredBindableInput,
) -> ReorderArrowAboveHoveredBindableResult {
    let hovered_bindable_id = input.hovered_bindable_id.clone().or_else(|| {
        input.point.and_then(|point| {
            input.bindables.as_ref().and_then(|bindables| {
                get_hovered_bindable(point, bindables, input.tolerance.unwrap_or(0.0))
                    .map(|bindable| bindable.id.clone())
            })
        })
    });

    let Some(hovered_bindable_id) = hovered_bindable_id else {
        return hovered_unchanged_result(&input.ordered_element_ids, None, Vec::new());
    };

    let anchor_lookup = input
        .anchor_element_ids_by_bindable_id
        .clone()
        .unwrap_or_else(HashMap::new);
    let anchor_element_ids = anchor_lookup
        .get(&hovered_bindable_id)
        .cloned()
        .unwrap_or_else(|| vec![hovered_bindable_id.clone()]);

    let reorder = reorder_arrow_above_elements(&ReorderArrowAboveElementsInput {
        ordered_element_ids: input.ordered_element_ids.clone(),
        arrow_id: input.arrow_id.clone(),
        anchor_element_ids: anchor_element_ids.clone(),
    });

    ReorderArrowAboveHoveredBindableResult {
        ordered_element_ids: reorder.ordered_element_ids,
        moved: reorder.moved,
        from_index: reorder.from_index,
        to_index: reorder.to_index,
        hovered_bindable_id: Some(hovered_bindable_id),
        anchor_element_ids,
    }
}

pub fn reordered_element_ids_from_hovered_reorder(
    result: &ReorderArrowAboveHoveredBindableResult,
) -> Option<Vec<String>> {
    result.moved.then(|| result.ordered_element_ids.clone())
}

pub fn did_engine_result_reorder(order_changed: Option<bool>) -> bool {
    order_changed.unwrap_or(false)
}

pub fn sort_bindables_by_z_index(bindables: &[BindableState]) -> Vec<BindableState> {
    let mut sorted = bindables.to_vec();
    sorted.sort_by(|left, right| {
        left.z_index
            .partial_cmp(&right.z_index)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    sorted
}
